cpp/src/arima/batched_kalman.cu

```diff
@@ -442,79 +631,135 @@ __global__ void batched_kalman_loglike_kernel(const double* d_vs,
   double* d_RQ = RQb.raw_data();
   const double* d_R = Rb.raw_data();
   thrust::for_each(thrust::cuda::par.on(stream), counting,
                    counting + batch_size, [=] __device__(int bid) {
                      double sigma2 = d_sigma2[bid];
-                     for (int i = 0; i < r; i++) {
-                       d_RQ[bid * r + i] = d_R[bid * r + i] * sigma2;
+                     for (int i = 0; i < rd; i++) {
+                       d_RQ[bid * rd + i] = d_R[bid * rd + i] * sigma2;
                      }
                    });
-  MLCommon::LinAlg::Batched::Matrix<double> RRT =
+  MLCommon::LinAlg::Batched::Matrix<double> RQR =
     MLCommon::LinAlg::Batched::b_gemm(RQb, Rb, false, true);
 
   // Durbin Koopman "Time Series Analysis" pg 138
   ML::PUSH_RANGE("Init P");
-  MLCommon::LinAlg::Batched::Matrix<double> P =
-    MLCommon::LinAlg::Batched::b_lyapunov(Tb, RRT);
+  MLCommon::LinAlg::Batched::Matrix<double> P(rd, rd, batch_size, cublasHandle,
+                                              allocator, stream, true);
+  {
+    double* d_P = P.raw_data();
+
+    if (n_diff > 0) {
+      // Initialize the diffuse part with a large variance
+      /// TODO: pass this as a parameter
+      constexpr double kappa = 1e6;
+      thrust::for_each(thrust::cuda::par.on(stream), counting,
+                       counting + batch_size, [=] __device__(int bid) {
+                         double* b_P = d_P + rd * rd * bid;
+                         for (int i = 0; i < n_diff; i++) {
+                           b_P[(rd + 1) * i] = kappa;
+                         }
+                       });
+
+      // Initialize the stationary part by solving a Lyapunov equation
+      /// TODO: reduce amount of memory copies
+      MLCommon::LinAlg::Batched::Matrix<double> Ts =
+        MLCommon::LinAlg::Batched::b_2dcopy(Tb, n_diff, n_diff, r, r);
+      MLCommon::LinAlg::Batched::Matrix<double> RQRs =
+        MLCommon::LinAlg::Batched::b_2dcopy(RQR, n_diff, n_diff, r, r);
+      MLCommon::LinAlg::Batched::Matrix<double> Ps =
+        MLCommon::LinAlg::Batched::b_lyapunov(Ts, RQRs);
+      MLCommon::LinAlg::Batched::b_2dcopy(Ps, P, 0, 0, r, r, n_diff, n_diff);
+    } else {
+      // Initialize by solving a Lyapunov equation
+      /// TODO: avoid copy
+      P = MLCommon::LinAlg::Batched::b_lyapunov(Tb, RQR);
+    }
+  }
   ML::POP_RANGE();
 
-  // Initialize the state alpha as the solution of (I - T) x = c
-  // Note: optimized as c = (mu 0 ... 0)'
+  // Initialize the state alpha by solving (I - T*) x* = c with:
+  //     | mu |
+  // c = | 0  |
+  //     | .  |
+  //     | 0  |
+  // T* = T[d+s*D:, d+s*D:]
+  // x* = alpha_0[d+s*D:]
   MLCommon::LinAlg::Batched::Matrix<double> alpha(
-    r, 1, batch_size, handle.getImpl().getCublasHandle(),
+    rd, 1, batch_size, handle.getImpl().getCublasHandle(),
     handle.getDeviceAllocator(), stream, true);
+  /// TODO: remove memset and set to 0 only the elements that need to
```

Comment:
is this a relevant comment? I don't see any memset happening here.

Reply:
The memset happens when initializing a batched matrix with `setZero=true`. I'll address these _TODO_ items